lib/teleterm/services/clientcache/clientcache.go

```diff
@@ -0,0 +1,182 @@
+// Teleport
+// Copyright (C) 2024 Gravitational, Inc.
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package clientcache
+
+import (
+	"context"
+	"sync"
+
+	"github.com/gravitational/trace"
+	"github.com/sirupsen/logrus"
+	"golang.org/x/sync/singleflight"
+
+	"github.com/gravitational/teleport/lib/client"
+	"github.com/gravitational/teleport/lib/teleterm/api/uri"
+	"github.com/gravitational/teleport/lib/teleterm/clusters"
+)
+
+// Cache stores clients keyed by cluster URI.
+// Safe for concurrent access.
+// Closes all clients and wipes the cache on Clear.
+type Cache struct {
+	clusters.Resolver
```

Comment:
I didn't notice that before, but I'd say that embedding one struct within another is rarely needed. Unless we want other callsites to be able to call `cache.ResolveCluster`?

In case of Cache and Config, I think it'd be preferable for Config to have Resolver and Log fields. Then Cache itself can have `cfg Config` field and refer to Resolver and Log through cfg, e.g. `c.cfg.Resolver`.

This way we don't have to keep expanding Cache when we add fields to Config.